File: Easy/0202_happy-number/happy-number.go
package main

import (
	"fmt"
)

func sumSquares(n int) int {
	sum := 0
	for n > 0 {
		d := n % 10
		sum += d * d
		n /= 10
	}
	return sum
}

func isHappy(n int) bool {
	slow, fast := n, sumSquares(n)

	for fast != 1 && slow != fast {
		slow = sumSquares(slow)
		fast = sumSquares(sumSquares(fast))
	}
	return fast == 1
}

func main() {
	tests := []int{
		19,
    2,
    1,
    10,
    42,
    2147483647,
	}

	for _, n := range tests {
		fmt.Printf("---\nN: %d, isHappy: %t\n", n, isHappy(n))
	}
}

File: Easy/0202_happy-number/happy-number.py
#!/bin/python3

from typing import List

class Solution:
    def isHappy(self, n: int) -> bool:
        seen = set()
        while n != 1 and n not in seen:
            seen.add(n)
            n = sum(int(digit) ** 2 for digit in str(n))
        return n == 1

def main():
    tests : List[int] = [
        19,
        2,
        1,
        10,
        42,
        2147483647,
    ]

    for n in tests:
        print(f"---\nN: {n} isHappy: {Solution().isHappy(n)}")

if __name__ == "__main__":
    main()
